docs/testing-strategy.md

```diff
@@ -0,0 +1,152 @@
+the perspective of someone using the application. This ensures we're testing
+the system in a way representative of expected user behaviour. This makes the
+tests some of the most valuable of the application, however they are slower
+than other tests to run and can be difficult to debug. Therefore they are not
+intended to exhaustively test all the scenarios that can occur.
+
+Characteristics of feature tests:
+
+- they should test across multiple requests - consider a
+  [request test](#request-tests) to test a single request;
+- they should navigate the application using the same means as a user, for
+  example clicking links or submitting forms;
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
+  can be tested at a higher granularity in [unit](#unit-tests) or
+  [request](#request-tests) tests;
+- they shouldn't test the effects a test has on the database, only user visible
+  signs of success should be asserted on;
+- they shouldn't mock application code;
+- there may be mocking and assertions that particular API calls are made to
+  external services;
+- they should be written in the [readable feature test
+  style][readable-feature-tests] popularised by Futurelearn.
+
+## Request tests
+
+The purpose of these tests is to test the HTTP API for the application from
+a machine perspective. They are used to determine that for a given input and
+application state a particular response is returned. These differ from feature
+tests in that they test a single endpoint at a time and can exhaust the logical
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
+- they should assert against the HTTP responses involved in the request, and
+  not side-effects such as database changes, these are better suited to [unit
+  testing](#unit-tests) an [interactor][interactors];
+- they should focus on the effects of a single endpoint with a single HTTP
+  method at a time, for example `POST /documents`, for multiple endpoints
+  consider a [feature test](#feature-tests);
+- they act as integration tests and thus shouldn't mock application code;
+- they can mock external API calls, but these should not be asserted against -
+  this is better suited to [unit tests](#unit-tests);
+- they shouldn't be used to extensively test HTML output, this is better suited
+  by [view tests](#view-tests);
+- they may not be necessary for places where a response doesn't have logic
+  and already has coverage provided by a feature test.
```

Comment:
This brings up a question in my head on whether we should test in request tests the 'happy path' which is more than likely covered by a feature test. We are currently doing this for a few of the requests tests (withdraw and unwithdraw etc) but should this be a hard and fast rule?

Reply:
I'd think if a route has different responses it should have request tests. It doesn't sit quite right with me to skip one of them as that would be covered by a feature test (feels overly coupled).

Just seems a bit overkill to test very basic requests which are just show something. Does that make sense? Happy to try rephrase a bit if you have some ideas.

Reply:
That makes sense, I think I was coming at it from the point of if I'm creating a request test for a thing which can have multiple responses should I always cover the scenario in which the feature covers ('happy' scenario). This is answered by your first paragraph! 😄